Src/IronPython.Modules/_codecs.cs

```diff
@@ -172,69 +172,14 @@ public static PythonTuple ascii_encode(CodeContext context, string input, string
-                if (data[i] == '\\') {
-                    if (i == data.Count - 1) throw PythonOps.ValueError("\\ at end of string");
-
-                    switch ((char)data[++i]) {
-                        case 'a': res.Append((char)0x07); break;
-                        case 'b': res.Append((char)0x08); break;
-                        case 't': res.Append('\t'); break;
-                        case 'n': res.Append('\n'); break;
-                        case 'r': res.Append('\r'); break;
-                        case '\\': res.Append('\\'); break;
-                        case 'f': res.Append((char)0x0c); break;
-                        case 'v': res.Append((char)0x0b); break;
-                        case '\n': break;
-                        case 'x':
-                            if (++i < data.Count && CharToInt((char)data[i], out int dig1)
-                                    && ++i < data.Count && CharToInt((char)data[i], out int dig2)) {
-                                res.Append((char)(dig1 * 16 + dig2));
-                            } else {
-                                switch (errors) {
-                                    case "strict":
-                                        throw PythonOps.ValueError("invalid \\x escape at position {0}", i);
-                                    case "replace":
-                                        res.Append("?");
-                                        i--;
-                                        break;
-                                    default:
-                                        throw PythonOps.ValueError("decoding error; unknown error handling code: " + errors);
-                                }
-                            }
-                            break;
-                        default:
-                            res.Append("\\" + (char)data[i]);
-                            break;
-                    }
-                } else {
-                    res.Append((char)data[i]);
-                }
-
-            }
-            return PythonTuple.MakeTuple(Bytes.Make(res.ToString().MakeByteArray()), data.Count);
-        }
-
-        private static bool CharToInt(char ch, out int val) {
-            if (char.IsDigit(ch)) {
-                val = ch - '0';
-                return true;
-            }
-            ch = char.ToUpper(ch);
-            if (ch >= 'A' && ch <= 'F') {
-                val = ch - 'A' + 10;
-                return true;
-            }
-
-            val = 0;
-            return false;
+            var res1 = LiteralParser.ParseBytes(data, 0, data.Count, isRaw: false, normalizeLineEndings: false);
+            return PythonTuple.MakeTuple(Bytes.Make(res1.ToArray()), data.Count);
         }
 
-        public static PythonTuple/*!*/ escape_encode([BytesConversion]IList<byte> text, string errors = "strict") {
+        public static PythonTuple escape_encode([BytesConversion]IList<byte> text, string errors = "strict") {
```

Comment:
Just in case you didn't know already. The `/*!*/` comments are non-null annotations. Need to look into turning on C#8 nullable.

Reply:
I knew they were meant to indicate non-nullness, but I thought they are obsolete. The tool that was checking them is defunct and the project developing the tool seems abandoned. I removed the marker here for the sake of consistency: only `escape_encode` return type was marked with `/*!*/`, but all of the encoding and decoding methods in `_codecs` return the same non-null tuple and are not marked. Nevertheless, since you have plans for the marker, I will add it back in the next commit. Moving forward, I will leave them alone but also not introduce any new ones.

The idea of employing non-nullable reference types sounds appealing. Is there an "Issue" for it? We also have the `[NotNull]` attribute serving a similar (?) purpose.

Reply:
I'm fine with the leave it alone approach. At this point they're only useful to help with future nullable annotations.

I have not filed an issue for it. I was planning on trying to annotate a few files before opening it up (although if someone else beats me to it then great).

As for replacing the `NotNullAttribute` it might be possible, although I think the C#8 `NullableAttribute` is embedded and internal? I have not given it any thought at this point.